File: app/schemas/documents.py
"""Document ingestion API schemas."""

from typing import Literal

from pydantic import BaseModel, Field


class DocumentIngestResponse(BaseModel):
    """Response returned after a successful document ingestion."""

    document_id: str
    filename: str
    content_type: str
    file_type: str
    file_size: int = Field(ge=0)
    checksum: str
    source: str
    page_count: int = Field(ge=1)
    pages_stored: int = Field(ge=1)
    chunks_stored: int = Field(ge=1)
    status: Literal["ingested"] = "ingested"


class DocumentBatchIngestItem(BaseModel):
    """Single document result within a batch upload response."""

    document_id: str
    filename: str
    content_type: str
    file_type: str
    file_size: int = Field(ge=0)
    checksum: str
    source: str
    page_count: int = Field(ge=1)
    pages_stored: int = Field(ge=1)
    chunks_stored: int = Field(ge=1)
    status: Literal["ingested"] = "ingested"


class DocumentBatchIngestResponse(BaseModel):
    """Response returned after ingesting multiple documents."""

    documents: list[DocumentBatchIngestItem]
    total_documents: int = Field(ge=1)
    status: Literal["ingested"] = "ingested"


class DocumentDeleteResponse(BaseModel):
    """Response returned after deleting an ingested document."""

    document_id: str
    chunks_deleted: int = Field(ge=0)
    checksum: str | None = None
    filename: str | None = None
    status: Literal["deleted"] = "deleted"

File: frontend/src/lib/types/documents.ts
/** Matches `DocumentIngestResponse` from the backend. */
export interface DocumentIngestResponse {
  document_id: string;
  filename: string;
  content_type: string;
  file_type: string;
  file_size: number;
  checksum: string;
  source: string;
  page_count: number;
  pages_stored: number;
  chunks_stored: number;
  status: "ingested";
}

/** Matches `DocumentSummaryResponse` from the backend. */
export interface DocumentSummary extends DocumentIngestResponse {
  ingested_at: string | null;
}

/** Matches `DocumentListResponse` from the backend. */
export interface DocumentListResponse {
  documents: DocumentSummary[];
  total_documents: number;
  status: "ok";
}

/** Matches `DocumentBatchIngestItem` from the backend. */
export type DocumentBatchIngestItem = DocumentIngestResponse;

/** Matches `DocumentBatchIngestResponse` from the backend. */
export interface DocumentBatchIngestResponse {
  documents: DocumentBatchIngestItem[];
  total_documents: number;
  status: "ingested";
}

export type DocumentUploadResponse = DocumentIngestResponse | DocumentBatchIngestResponse;

export function isBatchUploadResponse(
  response: DocumentUploadResponse,
): response is DocumentBatchIngestResponse {
  return "documents" in response;
}

export function uploadResponseToDocuments(
  response: DocumentUploadResponse,
): DocumentIngestResponse[] {
  if (isBatchUploadResponse(response)) {
    return response.documents;
  }
  return [response];
}
